pkg/server/ca/validation.go

```diff
@@ -0,0 +1,62 @@
+package ca
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"fmt"
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+)
+
+var (
+	validationPubkey, _ = pemutil.ParsePublicKey([]byte(`-----BEGIN PUBLIC KEY-----
+MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEzLY1/SRlsMJExTnuvzBO292RjGjU
+3L8jFRtmQl0CjBeHdxUlGK1OkNLDYh0b6AW4siWt+y+DcbUAWNb14e5zWg==
+-----END PUBLIC KEY-----`))
+)
+
+type X509CAValidator struct {
+	TrustDomain spiffeid.TrustDomain
+	Signer      crypto.Signer
+}
+
+func (v X509CAValidator) ValidateX509CA(ctx context.Context, x509CA, upstreamRoots []*x509.Certificate) error {
+	params := X509SVIDParams{
+		SpiffeID:  v.TrustDomain.NewID("/spire/throwaway"),
+		PublicKey: validationPubkey,
+	}
+
+	var bundle *x509bundle.Bundle
+	var upstreamChain []*x509.Certificate
+	if len(upstreamRoots) > 0 {
+		// If there are upstream roots, this is an upstream signed intermediate
+		// CA, so set the bundle to the upstream roots and set the upstream
+		// chain appropriately.
+		bundle = x509bundle.FromX509Authorities(v.TrustDomain, upstreamRoots)
+		upstreamChain = x509CA
+	} else {
+		// If there are no upstream roots, this is a self-signed CA so set
+		// the bundle to the CA itself and don't set the upstream chain.
+		bundle = x509bundle.FromX509Authorities(v.TrustDomain, x509CA)
```

Comment:
If i'm no t missing somethings looks like we are not creating a validator in case of selfsigning certs, is it added for precaution? and if we are not creating validator for selfsigning, and we get here, what may we do?

Reply:
I made the validator able to handle the case but didn't actually do the validation on self-signed CAs (since the original issue for the PR was about upstream authority). I'm torn on whether or not we should do it though. Since the code to generate the self-signed code is all controlled internally, seems like we can get sufficient coverage through unit-testing and not rely on a runtime check...